### PR TITLE
*: make union execution parallel and fix a union plan bug.

### PR DESCRIPTION
**Pull request by hanfei1991**
An enhancement and a bug fix
1. make union execution parallel
2. order can't push across union / unionall but limit can.
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu @winoros @lamxTyler PTAL

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1907,51 +1907,75 @@ func (e *TrimExec) Next() (*Row, error) {
...
-				var val types.Datum
-				val, err = row.Data[i].ConvertTo(e.ctx.GetSessionVars().StmtCtx, col.RetType)
-				if err != nil {
-					return nil, errors.Trace(err)
-				}
-				row.Data[i] = val
-			}
+			return
 		}
+		e.rowsCh <- row
```
Comment (shenli) [inline executor/executor.go]: If there are too many rows, the overhead of channel will be huge. It is better to use a batch way.

File: plan/physical_plan_builder.go [hunk 2]
```diff
@@ -871,24 +871,16 @@ func (p *Union) convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanInfo,
...
-	var count uint64
 	for _, child := range p.GetChildren() {
-		newProp := convertLimitOffsetToCount(prop)
-		newProp.props = make([]*columnProp, 0, len(prop.props))
-		for _, c := range prop.props {
-			idx := p.GetSchema().GetIndex(c.col)
-			newProp.props = append(newProp.props, &columnProp{col: child.GetSchema()[idx], desc: c.desc})
-		}
+		newProp := limitProperty(limit)
 		info, err = child.(LogicalPlan).convert2PhysicalPlan(newProp)
```
Comment (coocood) [inline plan/physical_plan_builder.go]: `info` is the returned variable, should not be used as a temporary variable.

File: plan/physical_plan_builder.go [hunk 3]
```diff
@@ -871,24 +871,16 @@ func (p *Union) convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanInfo,
...
 	childInfos := make([]*physicalPlanInfo, 0, len(p.children))
-	var count uint64
 	for _, child := range p.GetChildren() {
-		newProp := convertLimitOffsetToCount(prop)
-		newProp.props = make([]*columnProp, 0, len(prop.props))
-		for _, c := range prop.props {
-			idx := p.GetSchema().GetIndex(c.col)
-			newProp.props = append(newProp.props, &columnProp{col: child.GetSchema()[idx], desc: c.desc})
-		}
+		newProp := limitProperty(limit)
```
Comment (coocood) [inline plan/physical_plan_builder.go]: Does this mean "limit" can be pushed to children without "order by" is pushed?
What about this case:
```
create t1 (a int);
insert t1 values (2), (1);
create t2 (a int);
insert t2 values (3), (4);
select * from t1 union all (select * from t2) order by a limit 1;
```
The correct result is `1`, but if limit is pushed to children, the result would be `2`.

**Comment by coocood** [comment]
LGTM

File: executor/executor.go [hunk 4]
```diff
@@ -1907,51 +1907,99 @@ func (e *TrimExec) Next() (*Row, error) {
...
+		}
+		go e.waitAllFinished()
+		e.inited = true
+	}
+	if e.cursor >= len(e.rows) {
+		var rows []*Row
+		var err error
+		select {
+		case rows, _ = <-e.rowsCh:
+		case err, _ = <-e.errCh:
```
Comment (shenli) [inline executor/executor.go]: If we meet error here, should we call e.wg.Done() for len(e.Srcs) time? Or go e.waitAllFinished() will leak.

File: executor/executor.go [hunk 5]
```diff
@@ -1907,51 +1907,99 @@ func (e *TrimExec) Next() (*Row, error) {
...
+	e.cursor++
+	return row, nil
 }
 
 // Close implements the Executor Close interface.
 func (e *UnionExec) Close() error {
+	e.finished.Store(true)
+	<-e.closedCh
 	e.cursor = 0
+	e.inited = false
```
Comment (shenli) [inline executor/executor.go]: e.rows = nil

File: executor/executor.go (hunk 4 above)
Comment (hanfei1991) [inline executor/executor.go]: When we meet error in fetch data, we will set finished be true and notify all goroutines to stop. wg.Done is added in defer for every fetch data goroutines.

File: executor/executor.go [hunk 6]
```diff
@@ -1907,51 +1907,99 @@ func (e *TrimExec) Next() (*Row, error) {
...
+}
+
 // Next implements the Executor Next interface.
 func (e *UnionExec) Next() (*Row, error) {
-	for {
-		if e.cursor >= len(e.Srcs) {
-			return nil, nil
+	if !e.inited {
+		e.finished.Store(false)
+		e.rowsCh = make(chan []*Row, batchSize*len(e.Srcs))
```
Comment (hanfei1991) [inline executor/executor.go]: This part is ok, when the upstream executor process data every fast but the downstream one is slow
a large buffer benifits a lot.

File: executor/executor.go [hunk 7]
```diff
@@ -1907,51 +1907,100 @@ func (e *TrimExec) Next() (*Row, error) {
...
-		row, err := sel.Next()
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		if row == nil {
-			e.cursor++
-			continue
-		}
-		if e.cursor != 0 {
-			for i := range row.Data {
```
Comment (shenli) [inline executor/executor.go]: You should keep this logic.

File: executor/executor.go [hunk 8]
```diff
@@ -1907,51 +1907,99 @@ func (e *TrimExec) Next() (*Row, error) {
...
 
+func (e *UnionExec) waitAllFinished() {
+	e.wg.Wait()
+	close(e.rowsCh)
+	close(e.closedCh)
+}
+
+func (e *UnionExec) fetchData(idx int) {
+	defer func() {
+		e.wg.Done()
```
Comment (zimulala) [inline executor/executor.go]: Use `defer e.wg.Done()` directly.

File: executor/executor.go [hunk 9]
```diff
@@ -1907,51 +1907,98 @@ func (e *TrimExec) Next() (*Row, error) {
...
+			row, err := e.Srcs[idx].Next()
+			if err != nil {
+				e.finished.Store(true)
+				e.errCh <- err
+				return
+			}
+			if row == nil {
+				if len(rows) > 0 {
+					e.rowsCh <- rows
+				}
```
Comment (zimulala) [inline executor/executor.go]: If `e.row == nil` and `len(rows) == 0` are true,  `e.rowCh` and `e.errCh` are all empty, the `select` in line 1977 is blocked.

Comment (hanfei1991) [inline executor/executor.go]: It doesn't matter.

File: executor/executor.go [hunk 10]
```diff
@@ -1907,51 +1907,109 @@ func (e *TrimExec) Next() (*Row, error) {
...
 // Next implements the Executor Next interface.
 func (e *UnionExec) Next() (*Row, error) {
-	for {
-		if e.cursor >= len(e.Srcs) {
-			return nil, nil
+	if !e.inited {
+		e.finished.Store(false)
+		e.rowsCh = make(chan []*Row, batchSize*len(e.Srcs))
+		e.errCh = make(chan error, len(e.Srcs))
+		e.closedCh = make(chan struct{}, 1)
```
Comment (zimulala) [inline executor/executor.go]: Don't need `1` here.

**Comment by zimulala** [comment]
LGTM

File: executor/executor.go [hunk 11]
```diff
@@ -1907,51 +1907,109 @@ func (e *TrimExec) Next() (*Row, error) {
...
+				return
+			}
+			if row == nil {
+				if len(rows) > 0 {
+					e.rowsCh <- rows
+				}
+				return
+			}
+			// TODO: Add cast function in plan building phase.
+			for j := range row.Data {
```
Comment (shenli) [inline executor/executor.go]: If idx == 0, we do not need to convert data.

Comment (shenli) [inline executor/executor.go]: See Line 1936 in the master code.

File: executor/executor.go (hunk 6 above)
Comment (shenli) [inline executor/executor.go]: OK

File: executor/executor.go (hunk 4 above)
Comment (shenli) [inline executor/executor.go]: Got it.

**Comment by shenli** [comment]
Rest LGTM

**Pull request merged**